docs/cndb_omf.md

```diff
@@ -0,0 +1,294 @@
+described by the TXC record.
+
+- TX id: Transaction id to tie this record to the corresponding TX record.
+- Cnid: (CN id) A unique identifier for the KVS.
+- Node id: Node id to which this kvset belongs. Node ids are found in TXNODE records.
+- Dgen: Data generation number. Everytime a new kvset is created during ingest, it gets a new dgen. Kvsets
+  resulting from a compaction operation adopt the dgen of the src kvset with the largest dgen.
+- Vused:
+- Compc: Number of times this kvset has undergone compactions.
+- Scatter:
+
+### TXD
+- TX id: Transaction id to tie this record to the corresponding TX record.
+- Cnid: (CN id) A unique identifier for the KVS.
+- Tag: A unique number used to match with corresponding ACK-C records.
+- Number of mblocks.
+- List of Mblock IDs
+
+### ACK
+All records until now are meant to log an intent. `ACK` and `NAK` commit and abort transactions (or
+parts of transactions) repectively.
+
+Types of `ACK` records:
+1. `ACK-N`: Ack a `TXNODE` record
+2. `ACK-C`: Ack a `TXC` record. One `ACK-C` record for every `TXC` record.
+3. `ACK-D`: Ack a `TXD` record. One `ACK-D` record for every `TXD` record.
+
+- TX id: Transaction id to tie this record to the corresponding TX record.
+- Cnid: (CN id) A unique identifier for the KVS. Used only by Ack-D.
+- Tag: Must match the TXC/TXD record that is being Ack-ed.
+
+### NAK
+Mark txn as aborted.
+- TX id: Transaction id to tie this record to the corresponding TX record.
+
+How do you NAK a prtial update?!!
+
+## Operation
+
+### New KVDB and KVS
+- A new KVDB starts out with a CNDB that contains the version and a meta record.
+- When a new KVS is created, an info record is appended to the CNDB. When the record type is set to
+  `CNDB_TYPE_INFOD`, this same record marks a KVS delete operation.
+
+### Tree shape changes
+- On startup, CNDB will read through `TXNODE` records in order and maintain a `master node list` accordingly.
+- When a `TXNODE` record is read, an entry is added to a pending node list. When this txn is committed, the
+  `master node list` is updated with the node replacements.
+- On mdc compaction, CNDB would write its `TXNODE` record based on the `master node list`.
+- At replay too, the last two steps should give a new tree.
+- Operations:
+  - First root spill.
+  - Split and Join.
+  - CNDB compaction: Encapsulate the current tree shape in the compacted CNDB view.
+
+## Sample logs
+
+Part 1: Beginning.
+```
+# Create KVDB
+version magic=[...] version=01 captgt=1234567890
+meta seqno=01
+
+# Create KVS
+> info cnid=01 pfxlen=00 flags=0x0 name=kvs01 metasz 2176 ...
+
+# Ingest - no deletes
+> tx    txid=01 seqno=[...] ingest_id=[...] txhorizon=[...]
+> txc   txid=01 cnid=01 tag=01 kbk.cnt=01 vblk.cnt=01 ...
+> ack-C txid=01 cnid=01 tag=01
+
+# First root spill. Creates a single node at level 1 - node-01.
+> tx     txid=02 seqno=[...] ingest_id=CNDB_INVAL_INGESTID txhorizon=CNDB_INVAL_HORIZON
```

Comment:
Seems like we are shoehorning the record to also cover spills, which feels very dirty. Why not just have a different record type?

Reply:
Spill, split, join, k-compact and kv-compact are just different types of tree mutations (i.e., create and delete nodes and kvsets).  CNDB processes tree mutations and doesn't need to know if the mutation was a spill or something else.

Reply:
I'll give it some more thought.